### PR TITLE
feat(ui): Render 2000+ nodes DAG acceptably.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to USERS.md.

Address one part of #2367.

# Changes

* Replace the HTML based rendering with SVG rendering.
* Simplified the logic - removing extranous data from loops and the like.
* The filters now sit at the top left of the DAG.

# Questions for the Reviewers

* I have removed the arrow heads from the graph edges. Do we want these badly?
* Text rendering is a bit rough. Is it too bad?

# Screenshots
## Before
https://docs.google.com/document/d/1_hehmELNdFdffSlUw1TLH13AMikXxKshWrndH7KhxrQ/edit

### CODE REVIEW

File: ui/src/app/workflows/components/workflow-dag/workflow-dag.scss [hunk 1]
```diff
@@ -2,143 +2,66 @@
...
+  }
 
-        &.small {
-            background-color: transparent;
-            border: none;
-            padding-left: 0;
+  .label {
+    stroke: $argo-color-gray-7;
+    font-weight: lighter;
+  }
```
Comment (simster7) [inline ui/src/app/workflows/components/workflow-dag/workflow-dag.scss:32]: This stroke and font weight makes the font hard to read in my opinion. I would suggest removing it.

From
image

To:
image

**Comment by simster7** [comment]
Perhaps also consider allowing node labels to be longer? Or only show the suffix? Or have an option to do either? I have seen some complaints in the past that node labels are getting cut off too early, especially since the most significant information in a node name tends to be at the end.

**Comment by alexec** [comment]
> Perhaps also consider allowing node labels to be longer? Or only show the suffix? Or have an option to do either? I have seen some complaints in the past that node labels are getting cut off too early, especially since the most significant information in a node name tends to be at the end.

How about middle ellipsis?

**Comment by simster7** [comment]
> How about middle ellipsis?

I would support this

File: ui/src/app/workflows/components/workflow-dag/workflow-dag.scss (hunk 1 above)
Comment (alexec) [reply ui/src/app/workflows/components/workflow-dag/workflow-dag.scss]: Ok. So I've corrected the font used and increased the font size to 12pt.

**Comment by alexec** [comment]
Changes

* Fixed the clicking bug (optimize one thing, you break another).
* Increased the horizontal spacing - I'd compacted this.
* Render the labels in three ways so we ALWAY show more than before:
  * Short - single line?
  * Medium  - split onto two lines.
  * Long - split into two lines with ellipsis.

File: ui/src/models/workflows.ts [hunk 2]
```diff
@@ -916,15 +916,17 @@ export interface WorkflowStep {
...
 export const NODE_PHASE = {
     PENDING: 'Pending',
     RUNNING: 'Running',
     SUCCEEDED: 'Succeeded',
     SKIPPED: 'Skipped',
     FAILED: 'Failed',
-    ERROR: 'Error'
+    ERROR: 'Error',
+    // this is a fake phase - no actual node even has this
+    SUSPENDED: 'Suspended'
```
Comment (simster7) [inline ui/src/models/workflows.ts:929]: Minor:

Could we add a comment here as to why this is needed? Something like "this is used to be able to filter suspended nodes in the DAG view". Might make is easier to tell why this exists only in the UI and not in the back-end code

**Comment by alexec** [comment]
> * The white space in a node icon is not recognized as part of the node for a click. This can make some icons that have a lot of while harder to click. See: https://cl.ly/b0b71d1ff570

Fixed - circle, icon, and text are all clickable now.

> * The "Pending" (yellow clock) node isn't animated. Gives the impression that there was something wrong if you're used to the animation.

I think that only when something is happening should it be animated - nothing is happening here - which is why I changed it. Thoughts?

> * I still can't see arrows on my browser (see screenshot). I am using Firefox 76.0. Let me know if you need more details/help in debugging this

I had an incorrect path, but only FF failed to render it. We def want FF support:

Screen Shot 2020-05-07 at 10 00 36 AM

> Could we add a comment here as to why this is needed? Something like "this is used to be able to filter suspended nodes in the DAG view". Might make is easier to tell why this exists only in the UI and not in the back-end code

I've removed this - it was just wrong - there is not such phase as "suspended" - just a flag.

**Comment by simster7** [comment]
Arrows are back, but now nodes are disappearing, seemingly arbitrarily:

image

> I think that only when something is happening should it be animated - nothing is happening here - which is why I changed it. Thoughts?

I think "Pending" is akin to "Waiting" and would make sense to show it as animated. Not super important either way though. Just mentioning because it's a change of what behavior we have now.

**Comment by alexec** [comment]
@simster7 

Can I get you to give me the full YAML of this problematic workflow - including status - which I can load into my dev env?

**Comment by simster7** [comment]
Sure thing:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  creationTimestamp: "2020-05-07T19:05:02Z"
  generateName: dag-diamond-
  generation: 11
  labels:
    workflows.argoproj.io/completed: "true"
    workflows.argoproj.io/phase: Succeeded
  name: dag-diamond-vzlxb
  namespace: argo
  resourceVersion: "12718"
  selfLink: /apis/argoproj.io/v1alpha1/namespaces/argo/workflows/dag-diamond-vzlxb
  uid: bb4a586b-6381-4537-86e3-ba5366a5362c
spec:
  arguments: {}
  entrypoint: diamond
  templates:
  - arguments: {}
    dag:
      tasks:
      - arguments:
          parameters:
          - name: message
            value: A
        name: A
        template: echo
      - arguments:
          parameters:
          - name: message
            value: B
        dependencies:
        - A
        name: B
        template: echo
      - arguments:
          parameters:
          - name: message
            value: C
        dependencies:
        - A
        name: C
        template: echo
      - arguments:
          parameters:
          - name: message
            value: D
        dependencies:
        - B
        - C
        name: D
        template: echo
    inputs: {}
    metadata: {}
    name: diamond
    outputs: {}
  - arguments: {}
    container:
      command:
      - echo
      - '{{inputs.parameters.message}}'
      image: alpine:3.7
      name: ""
      resources: {}
    inputs:
      parameters:
      - name: message
    metadata: {}
    name: echo
    outputs: {}
status:
  conditions:
  - status: "True"
    type: Completed
  finishedAt: "2020-05-07T19:05:12Z"
  offloadNodeStatusVersion: fnv:4269119374
  phase: Succeeded
  resourcesDuration:
    cpu: 5
    memory: 0
  startedAt: "2020-05-07T19:05:02Z"
```

**Comment by alexec** [comment]
Can I please have this with the nodes?

**Comment by simster7** [comment]
Edited the original comment to provide a different, but similar WF that suffers from the same problem

**Comment by alexec** [comment]
That was a straight out bug - I changed how I padded the SVG, but did not add enough padding to draw the bottom most and right most nodes.

**Comment by simster7** [comment]
Let me know when you'd like another review on this

File: test/e2e/expectedfailures/dag-disable-failFast.yaml [hunk 3]
```diff
@@ -7,7 +7,7 @@ spec:
   templates:
   - name: a
     container:
-      image:  argoproj/argosay:v1
```
Comment (alexec) [inline test/e2e/expectedfailures/dag-disable-failFast.yaml:10]: need to revert these

File: test/e2e/fixtures/given.go [hunk 4]
```diff
@@ -76,6 +76,7 @@ func (g *Given) checkImages(templates []wfv1.Template) {
 	imageWhitelist := map[string]bool{
 		"argoexec:" + imageTag: true,
 		"argoproj/argosay:v1":  true,
+		"argoproj/argosay:v2":  true,
```
Comment (alexec) [inline test/e2e/fixtures/given.go:79]: tech debt: create a new PR at some point to remove v1 completly

File: ui/src/app/shared/components/resource-submit.tsx [hunk 5]
```diff
@@ -72,10 +67,10 @@ export class ResourceSubmit<T> extends React.Component<ResourceSubmitProps<T>, R
                                     onChange={e => {
                                         try {
                                             this.readFile(e.target.files, (newResource: string) => {
-                                                const resource: T = jsYaml.load(e.currentTarget.value);
```
Comment (alexec) [inline ui/src/app/shared/components/resource-submit.tsx:75]: bug - did not work?

File: ui/src/app/shared/duration.ts [hunk 6]
```diff
@@ -28,5 +28,8 @@ export function formatDuration(seconds: number) {
 }
 
 export function wfDuration(status: models.WorkflowStatus) {
+    if (!status.startedAt) {
```
Comment (alexec) [inline ui/src/app/shared/duration.ts:31]: bug fix

File: ui/src/app/workflows/components/workflows-list/workflows-list.tsx [hunk 7]
```diff
@@ -120,6 +120,7 @@ export class WorkflowsList extends BasePage<RouteComponentProps<any>, State> {
                                     if (!wfValue || !wfValue.metadata) {
                                         return {valid: false, message: 'Invalid Workflow: metadata cannot be blank'};
                                     }
+                                    wfValue.metadata.namespace = wfValue.metadata.namespace || this.state.namespace;
```
Comment (alexec) [inline ui/src/app/workflows/components/workflows-list/workflows-list.tsx:123]: default the namespace

**Comment by alexec** [comment]
@simster7 

- This PR is now pork-barreled - I plan to extract the `argosay:v2` into another PR, so ignore that for the time being. 
- I've now gotten rid of Dagre and render using hand-coded coffman Graham algo. I'm not sure I'd even try to understand how this works.

File: ui/src/app/workflows/components/workflow-dag/graph/coffman-graham-sorter.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
+import {DfsSorter} from './dfs-sorter';
+import {Graph, node} from './graph';
+
+export class CoffmanGrahamSorter {
```
[the remark this answers is not in the document]
Comment (alexec) [reply ui/src/app/workflows/components/workflow-dag/graph/coffman-graham-sorter.ts:4]: it's not perfect, not as aesthetically pleasing as Network Simplex, but scales better

**Pull request merged**